lib/SILOptimizer/Transforms/ArrayElementValuePropagation.cpp

```diff
@@ -167,18 +191,51 @@ bool ArrayAllocation::isInitializationWithKnownElements() {
 ///
 /// We have to prove that the array value is not changed in between the
 /// creation and the method call to get_element.
-bool ArrayAllocation::findValueReplacements() {
+bool ArrayAllocation::findReplacements() {
   if (!isInitializationWithKnownElements())
     return false;
 
   // The array value was stored or has escaped.
   if (!analyseArrayValueUses())
     return false;
 
-  // No count users.
-  if (GetElementCalls.empty())
+  if (GetElementCalls.empty() && AppendContentsOfCalls.empty())
     return false;
 
+  if (!AppendContentsOfCalls.empty()) {
```

Comment:
You can do an early exit here that cleans up the control flow making it easier to read by reducing the level of indentation of most of the code here. Specifically:

if (AppendContentsOfCalls.empty()) {
   return collectForwardableValues();
}
LARGE BLOCK OF CODE
return collectForwardableValues();

Most importantly it communicates your intent that there is no more work in the no append contents case even after the if block.